Sources/SwiftParser/Specifiers.swift

```diff
@@ -0,0 +1,290 @@
+      unexpectedBeforeAsync.append(contentsOf: unexpected?.elements ?? [])
+      asyncKeyword = async
+    }
+
+    while let misspelledThrows = self.consume(ifAny: S.misspelledThrowsTokenKinds + S.misspelledAsyncTokenKinds) {
+      unexpectedBeforeThrows.append(RawSyntax(misspelledThrows))
+      if throwsKeyword == nil, S.misspelledThrowsTokenKinds.contains(misspelledThrows.tokenKind) {
+        // Let's synthesize a missing 'throws'. If we find a real throws specifier
+        // later, we will replace the missing token by the present token.
+        throwsKeyword = missingToken(.keyword(.throws))
+      }
+    }
+
+    if let handle = self.canRecoverTo(any: S.correctThrowsTokenKinds) {
+      let (unexpected, throwsKw) = self.eat(handle)
+      unexpectedBeforeThrows.append(contentsOf: unexpected?.elements ?? [])
+      throwsKeyword = throwsKw
+    }
+
+    while let misplacedSpecifier = self.consume(ifAnyIn: EffectSpecifier.self) {
+      if asyncKeyword == nil, S.correctAsyncTokenKinds.contains(misplacedSpecifier.tokenKind) || S.misspelledAsyncTokenKinds.contains(misplacedSpecifier.tokenKind) {
+        // Handle `async` after `throws`
+        asyncKeyword = missingToken(.keyword(.async))
+      }
+      unexpectedAfterThrows.append(RawSyntax(misplacedSpecifier))
+    }
+
+    if unexpectedBeforeAsync.isEmpty && asyncKeyword == nil && unexpectedBeforeThrows.isEmpty && throwsKeyword == nil && unexpectedAfterThrows.isEmpty {
+      return nil
+    }
+
+    return S(
+      RawUnexpectedNodesSyntax(unexpectedBeforeAsync, arena: self.arena),
+      asyncSpecifier: asyncKeyword,
+      RawUnexpectedNodesSyntax(unexpectedBeforeThrows, arena: self.arena),
+      throwsSpecifier: throwsKeyword,
+      RawUnexpectedNodesSyntax(unexpectedAfterThrows, arena: self.arena),
+      arena: self.arena
+    )
+  }
+
+  mutating func parseTypeEffectSpecifiers() -> RawTypeEffectSpecifiersSyntax? {
+    return parseEffectSpecifiers(RawTypeEffectSpecifiersSyntax.self)
+  }
+
+  mutating func parseDeclEffectSpecifiers() -> RawDeclEffectSpecifiersSyntax? {
+    return parseEffectSpecifiers(RawDeclEffectSpecifiersSyntax.self)
+  }
+
+  /// Consume any misplaced effect specifiers and return them in as unexpected tokens.
+  /// When a misplace effect specifier is consumed and `effectSpecifiers`
+  /// doesn't have  an effect specifier of that kind, modify `effectSpecifiers`
+  /// to have a missing specifier of that kind.
+  mutating func parseMisplacedEffectSpecifiers<S: RawEffectSpecifiersTrait>(_ effectSpecifiers: inout S?) -> RawUnexpectedNodesSyntax? {
+    var synthesizedAsync: RawTokenSyntax? = nil
+    var synthesizedThrows: RawTokenSyntax? = nil
+    var unexpected: [RawTokenSyntax] = []
+    while let specifier = self.consume(ifAnyIn: EffectSpecifier.self) {
+      unexpected.append(specifier)
+      if S.correctAsyncTokenKinds.contains(specifier.tokenKind) || S.misspelledAsyncTokenKinds.contains(specifier.tokenKind) {
```

Comment:
Could add a check for whether `effectSpecifiers` already has an async/throws to avoid both the missing token creation + creating a new node below when we don't actually need to.

Reply:
Good idea